File: android/app/src/main/java/com/openburnbar/data/insights/InsightCanvas.kt
package com.openburnbar.data.insights

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID

/**
 * A persistent dashboard composed of insight widgets.
 * Mirrors Swift InsightCanvas.
 */
@Serializable
data class InsightCanvas(
    val id: String = UUID.randomUUID().toString(),
    val title: String,
    val summary: String? = null,
    val symbolName: String = "sparkles.tv",
    val theme: InsightTheme = InsightTheme.AURORA,
    val widgets: List<InsightWidget> = emptyList(),
    val layout: InsightLayout = InsightLayout(),
    val filter: InsightFilter = InsightFilter(),
    val modelTag: InsightModelTag? = null,
    val schemaVersion: Int = CURRENT_SCHEMA_VERSION,
    val createdAt: String = "",
    val updatedAt: String = "",
    val lastRefreshedAt: String? = null,
    val origin: Origin = Origin.UserCreated,
    val sortIndex: Int = 0
) {
    @Serializable
    sealed class Origin {
        @Serializable data object UserCreated : Origin()
        @Serializable data class Template(val id: String) : Origin()
        @Serializable data class Composed(val prompt: String) : Origin()
        @Serializable data class Imported(val filename: String) : Origin()
    }

    companion object {
        const val CURRENT_SCHEMA_VERSION = 1
    }

    fun add(widget: InsightWidget): InsightCanvas {
        val newLayout = layout.placeNew(widget.id, widget.kind.defaultSpanColumns to widget.kind.defaultSpanRows)
        return copy(widgets = widgets + widget, layout = newLayout, updatedAt = nowISO())
    }

    fun remove(widgetID: String): InsightCanvas {
        return copy(
            widgets = widgets.filter { it.id != widgetID },
            layout = layout.remove(widgetID),
            updatedAt = nowISO()
        )
    }

    fun replace(widget: InsightWidget): InsightCanvas {
        val idx = widgets.indexOfFirst { it.id == widget.id }
        if (idx < 0) return this
        return copy(widgets = widgets.toMutableList().apply { this[idx] = widget }, updatedAt = nowISO())
    }

    /** Update a single widget in place via a mutation function. */
    fun update(widgetID: String, mutate: (InsightWidget) -> InsightWidget): InsightCanvas {
        val idx = widgets.indexOfFirst { it.id == widgetID }
        if (idx < 0) return this
        val updated = mutate(widgets[idx])
        return copy(widgets = widgets.toMutableList().apply { this[idx] = updated }, updatedAt = nowISO())
    }
}

private fun nowISO(): String = java.time.Instant.now().toString()

File: OpenBurnBarCore/Sources/OpenBurnBarInsights/SharedModels/InsightCanvas.swift
import Foundation

/// A persistent dashboard composed of insight widgets.
///
/// The user can have many canvases (saved boards) and switch between them
/// from the canvas library. Each canvas survives sync, export/import, and
/// schema evolution via `schemaVersion`.
public struct InsightCanvas: Codable, Hashable, Sendable, Identifiable {
    public static let currentSchemaVersion: Int = 1

    public let id: UUID
    public var title: String
    public var summary: String?
    public var symbolName: String
    public var theme: InsightTheme
    public var widgets: [InsightWidget]
    public var layout: InsightLayout
    public var filter: InsightFilter
    public var modelTag: InsightModelTag?
    public var schemaVersion: Int
    public var createdAt: Date
    public var updatedAt: Date
    public var lastRefreshedAt: Date?
    public var origin: Origin
    public var sortIndex: Int

    public init(
        id: UUID = UUID(),
        title: String,
        summary: String? = nil,
        symbolName: String = "sparkles.tv",
        theme: InsightTheme = .aurora,
        widgets: [InsightWidget] = [],
        layout: InsightLayout = InsightLayout(),
        filter: InsightFilter = InsightFilter(),
        modelTag: InsightModelTag? = nil,
        schemaVersion: Int = InsightCanvas.currentSchemaVersion,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        lastRefreshedAt: Date? = nil,
        origin: Origin = .userCreated,
        sortIndex: Int = 0
    ) {
        self.id = id
        self.title = title
        self.summary = summary
        self.symbolName = symbolName
        self.theme = theme
        self.widgets = widgets
        self.layout = layout
        self.filter = filter
        self.modelTag = modelTag
        self.schemaVersion = schemaVersion
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.lastRefreshedAt = lastRefreshedAt
        self.origin = origin
        self.sortIndex = sortIndex
    }

    public enum Origin: Codable, Hashable, Sendable {
        case userCreated
        case template(id: String)
        case composed(prompt: String)
        case imported(filename: String)
    }

    // MARK: - Mutations

    /// Add a widget and place it on the layout at the first available cell.
    public mutating func add(_ widget: InsightWidget) {
        widgets.append(widget)
        layout.placeNew(widgetID: widget.id, defaultSpan: widget.kind.defaultSpan)
        updatedAt = Date()
    }

    /// Remove a widget by id from both the array and the layout.
    public mutating func remove(widgetID: UUID) {
        widgets.removeAll { $0.id == widgetID }
        layout.remove(widgetID: widgetID)
        updatedAt = Date()
    }

    /// Replace a widget while preserving its identity and layout placement.
    public mutating func replace(_ widget: InsightWidget) {
        guard let idx = widgets.firstIndex(where: { $0.id == widget.id }) else { return }
        widgets[idx] = widget
        updatedAt = Date()
    }

    /// Update a single widget in place.
    public mutating func update(widgetID: UUID, _ mutate: (inout InsightWidget) -> Void) {
        guard let idx = widgets.firstIndex(where: { $0.id == widgetID }) else { return }
        var w = widgets[idx]
        mutate(&w)
        widgets[idx] = w
        updatedAt = Date()
    }
}
